File: two-pointers/0042-trapping-rain-water/solution.py
# https://leetcode.com/problems/trapping-rain-water/description/
from typing import List, Optional, Tuple, Dict, Set
from collections import defaultdict, deque, Counter, OrderedDict
from dataclasses import dataclass
import heapq
import bisect
import functools
import itertools
import math
import sys


class Solution:
    def solve(self, nums: list[int]) -> int:
      # For this problem, we'll do a squeezing two pointer, taking the lowest "edge"
      # at each point as the bounding wall, and shrinking the smaller of the two edges
      tot = 0
      n = len(nums)
      lp, rp = 0, n-1
      maxl, maxr = 0,0

      while lp < rp:
        left, right = nums[lp], nums[rp]
        divet = min(left,right)
        print(maxl, maxr, max(0, max(maxl,maxr) - divet))
        tot += max(0, max(maxl,maxr) - divet)
        if left < right:  # left smaller, shrink lp
          lp += 1
          maxl = max(maxl, left)
        else:
          rp -= 1
          maxr = max(maxr, right)

      return tot


@dataclass
class TestCase:
    nums: list[int]
    expected: int


if __name__ == "__main__":
    s = Solution()

    tests = [
        TestCase([0,1,0,2,1,0,1,3,2,1,2,1], 6),
        TestCase([4,2,0,3,2,5], 9),
    ]

    correct = 0
    for i, t in enumerate(tests):
        actual = s.solve(t.nums)
        passed = actual == t.expected
        correct += passed

        print(f"Test {i}: {'PASS' if passed else 'FAIL'}")
        print(f"\tExpected: {t.expected}")
        print(f"\tActual:   {actual}")

    print(f"\n{correct} / {len(tests)} correct.")

File: two-pointers/0042-trapping-rain-water/solution.cpp
#include <iostream>
#include <vector>

template <typename T>
std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << "[";
    for (size_t i = 0; i < v.size(); i++) {
        os << v[i];
        if (i + 1 < v.size()) os << ", ";
    }
    os << "]";
    return os;
}

class Solution {
public:
    int solve(std::vector<int>& nums) {
      size_t n = nums.size();
      int tot = 0;
      int lp=0, rp=n-1;
      int maxl=0, maxr=0;

      while (lp<rp) {
        int left=nums[lp], right=nums[rp];
        int divet = std::min(left,right);
        tot += std::max(0, std::max(maxl,maxr)-divet);
        if (left < right) {
          lp += 1;
          maxl = std::max(maxl, left);
        } else { // else right smaller
          rp -= 1;
          maxr = std::max(maxr, right);
        }
      }
      return tot;
    }
};

struct TestCase {
    std::vector<int> nums;
    int expected;
};

int main() {
    std::vector<TestCase> tests = {
      {{ 0,1,0,2,1,0,1,3,2,1,2,1 }, 6,},
      {{4,2,0,3,2,5}, 9,},
    };

    Solution sol;
    int correct = 0;

    for (size_t i = 0; i < tests.size(); i++) {
        TestCase& t = tests[i];
        auto actual = sol.solve(t.nums);
        bool passed = (actual == t.expected);
        correct += passed;

        std::cout << "Test " << i << ": " << (passed ? "PASS" : "FAIL") << "\n";
        std::cout << "\tExpected: " << t.expected << "\n";
        std::cout << "\tActual:   " << actual << "\n";
    }

    std::cout << "\n" << correct << " / " << tests.size() << " correct.\n";
    return 0;
}
